massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs

```diff
@@ -0,0 +1,38 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Test.Massiv.Core.Index.Ix where
+
+import Import
+
+import Data.Massiv.Core
+
+intToList :: Int -> [Int]
+intToList x =
+    if x >= 0
+        then [0 .. x - 1]
+        else reverse [x + 1 .. 0]
+
+instance GenUnchecked Ix0 where
+    genUnchecked = pure Ix0
+
+instance GenValid Ix0 where
+    genValid = pure Ix0
+
+instance GenUnchecked Ix2 where
+    genUnchecked = (:.) <$> genUnchecked <*> genUnchecked
+    shrinkUnchecked (a :. b) = [i :. j | i <- intToList a, j <- intToList b]
+
+instance GenValid Ix2 where
+    genValid = (:.) <$> genValid <*> genValid
```

Comment:
This can be generated

Reply:
It can, but the generated version is different from mine.
Generated: genValid = genUnchecked `shouldSatisfy` isValid
  where
    genUnchecked = (:.) <$> genUnchecked <*> genUnchecked
If "isValid" is not always true, my version will be faster.

Reply:
Ah right, fair point.